src/lib.rs

```diff
@@ -454,12 +502,53 @@ impl<'object, Endian> DebugInfo<'object, Endian>
                 // matches us better. We need to keep going until we find the best one.
             }
 
-            return Ok(Some((path, line, func.map(|u| u.0.name.to_string_lossy()))));
+            let func = func.map(|u| {
+                if unit.language.is_some() {
+                    debug_assert!(self.opts.with_demangling,
+                                  "We shouldn't even bother finding the DW_AT_language if we \
+                                   aren't demangling");
+                }
+                match unit.language {
+                    Some(gimli::DW_LANG_C_plus_plus) |
+                    Some(gimli::DW_LANG_C_plus_plus_03) |
+                    Some(gimli::DW_LANG_C_plus_plus_11) => demangle_cpp_symbol(u.0.name),
+                    Some(gimli::DW_LANG_Rust) => demangle_rust_symbol(u.0.name),
+                    _ => u.0.name.to_string_lossy(),
```

Comment:
It'd be nice if we could report this case to the user somehow. If they ask for demangling, and we don't support the language that's in use, we should report it. I'm not entirely sure what the best way to do so is though.

Also, it *could* be that some demangling is better than no demangling, even for unknown languages. Would it be better to fall back to `cpp` here?